### PR TITLE
Implement a Remote Backend

### PR DESCRIPTION
**Pull request by pieterlexis**
This PR adds a [Remote Backend](https://doc.powerdns.com/authoritative/backends/remote.html) 

ToDo

 - [x] Fix the tests
 - [ ] Documentation
 - [ ] LMDB backend for slave zones
   - [ ] rrset status for slave zones

### CODE REVIEW

File: yang/pdns-server.yang [hunk 1]
```diff
@@ -245,6 +254,111 @@ module pdns-server {
             when "../zonetype = 'slave'";
             min-elements 1;
         }
+
+        list rrset {
+            when "../zonetype = 'master' or ../zonetype = 'native'";
+            // Nearly a straight up copy from
+            // https://gitlab.labs.nic.cz/jetconf/jetconf-knot/blob/master/yang-modules/dns-zones-state@2017-01-12.yang
+            // RRSIG data is removed
```
Comment (Habbie) [inline yang/pdns-server.yang]: also remove NSEC* then?

File: src/remote-backend/getAllDomains.cc [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      zoneName = leaf->value_str();
+    }
+    if (leafName == "kind") {
+      kind = leaf->value_str();
+    }
+  }
+  if (!zoneName.empty()) {
+    ret["result"].push_back(makeDomainInfo(zoneName, kind));
+  }
+  sendResponse(response, ret);
```
Comment (Habbie) [inline src/remote-backend/getAllDomains.cc]: What about `id` and `serial`?

File: src/remote-backend/httpfunc.cc [hunk 3]
```diff
@@ -24,4 +24,24 @@ void RemoteBackend::notFound(const Rest::Request &request, Http::ResponseWriter
   sendError(response, string(), Http::Code::Not_Found);
 }
 
+std::string RemoteBackend::urlDecode(std::string& eString) {
+  // from http://www.cplusplus.com/forum/general/94849/
+  std::string ret;
```
Comment (Habbie) [inline src/remote-backend/httpfunc.cc]: `ret.reserve(eString.length())` ?

File: src/remote-backend/httpfunc.cc [hunk 4]
```diff
@@ -24,4 +24,24 @@ void RemoteBackend::notFound(const Rest::Request &request, Http::ResponseWriter
...
+  char ch;
+  size_t i;
+  unsigned int j;
+  for (i = 0; i < eString.length(); i++) {
+    if (int(eString[i]) == 37) {
+      sscanf(eString.substr(i + 1, 2).c_str(), "%x", &j);
```
Comment (Habbie) [inline src/remote-backend/httpfunc.cc]: If `eString` ends in `%8`, this will put `8` in `j`

Comment (Habbie) [reply src/remote-backend/httpfunc.cc]: `%xy` (i.e. no valid hex) will leave `j` silently unchanged

File: src/remote-backend/domain-id.cc [hunk 5]
```diff
@@ -0,0 +1,31 @@
+#include "remote-backend.hh"
+
+namespace pdns_sysrepo::remote_backend
+{
+uint32_t getNewId(const boost::bimap<std::string, uint32_t> &map) {
+  if (map.empty()) {
+    return 1;
+  }
+  uint32_t highest = map.left.rbegin()->second;
```
Comment (Habbie) [inline src/remote-backend/domain-id.cc]: Shouldn't this be `map.right.rbegin()->first`?

File: LICENSE [hunk 6]
```diff
@@ -1,4 +1,4 @@
-Copyright 2019 Pieter Lexis <pieter.lexis@powerdns.com>
+Copyright 2019-2020 Pieter Lexis <pieter.lexis@powerdns.com>
```
Comment (Habbie) [inline LICENSE]: remove the copyright year?

File: src/remote-backend/find-zone.cc [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    string token = str.substr(prev, pos - prev);
+    if (!token.empty())
+      tokens.push_back(token);
+    prev = pos + delim.length();
+  } while (pos < str.length() && prev < str.length());
+  return tokens;
+}
+
+string RemoteBackend::findBestZone(const string& qname) {
+  // I miss dnsname.cc :(
```
[the remark this answers is not in the document]
Comment (pieterlexis) [reply src/remote-backend/find-zone.cc]: Perhaps later :)

File: src/pdns-sysrepo.cc [hunk 8]
```diff
@@ -124,6 +126,12 @@ int main(int argc, char* argv[]) {
       spdlog::trace("signalhandlers registered, notifying systemd we're ready");
 
       sd_notify(0, "READY=1");
+      spdlog::info("Starting remote backend webserver");
+      
+      Pistache::Address a = "127.0.0.1:9100";
```
[the remark this answers is not in the document]
Comment (pieterlexis) [reply src/pdns-sysrepo.cc]: It is only used internally between pdns-sysrepo and pdns_server, but in the end, yes (#8).

File: src/remote-backend/find-zone.cc [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+#include <fmt/format.h>
+
+#include "remote-backend.hh"
+#include <spdlog/spdlog.h>
+
+namespace pdns_sysrepo::remote_backend {
+
+using std::vector;
+
+vector<string> split(const string& str, const string& delim) {
```
[the remark this answers is not in the document]
Comment (pieterlexis) [reply src/remote-backend/find-zone.cc]: :+1:

File: src/remote-backend/find-zone.cc [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+string RemoteBackend::findBestZone(const string& qname) {
+  // I miss dnsname.cc :(
+  auto labels = split(qname, ".");
+  string ret = qname;
+
+  auto tree = d_session->get_subtree("/pdns-server:zones");
+
+  // haha, no root support
```
[the remark this answers is not in the document]
Comment (pieterlexis) [reply src/remote-backend/find-zone.cc]: This comment is wrong :)

File: src/remote-backend/httpfunc.cc [hunk 11]
```diff
@@ -0,0 +1,27 @@
+#include "remote-backend.hh"
+
+namespace pdns_sysrepo::remote_backend
+{
+void RemoteBackend::sendError(Http::ResponseWriter& response, const string& error, const Http::Code& code) {
```
[the remark this answers is not in the document]
Comment (pieterlexis) [reply src/remote-backend/httpfunc.cc]: this is what clang format does :)

File: src/remote-backend/httpfunc.cc (hunk 4 above)
Comment (pieterlexis) [reply src/remote-backend/httpfunc.cc]: As the pdns_server is trusted, this is fine for now

File: src/remote-backend/domain-id.cc (hunk 5 above)
Comment (pieterlexis) [reply src/remote-backend/domain-id.cc]: This works in the tests. The map.left is <string, uint32_t>.

**Pull request merged**